File: go/food-chain/food_chain.go
package foodchain

import (
	"fmt"
	"strings"
)

// Define the animals and their corresponding special lines
var animals = []string{
	"", // Index 0 is empty to align with verse numbers starting at 1
	"fly",
	"spider",
	"bird",
	"cat",
	"dog",
	"goat",
	"cow",
	"horse",
}

var specialLines = []string{
	"", // Index 0 is empty to align with verse numbers starting at 1
	"I don't know why she swallowed the fly. Perhaps she'll die.",
	"It wriggled and jiggled and tickled inside her.",
	"How absurd to swallow a bird!",
	"Imagine that, to swallow a cat!",
	"What a hog, to swallow a dog!",
	"Just opened her throat and swallowed a goat!",
	"I don't know how she swallowed a cow!",
	"She's dead, of course!",
}

// Verse returns the verse for the given verse number
func Verse(v int) string {
	if v < 1 || v > 8 {
		return ""
	}

	var lines []string
	
	// First line is always the same pattern
	lines = append(lines, fmt.Sprintf("I know an old lady who swallowed a %s.", animals[v]))
	
	// Add the special line for this animal
	lines = append(lines, specialLines[v])
	
	// For the horse (verse 8), we're done after the special line
	if v == 8 {
		return strings.Join(lines, "\n")
	}
	
	// For verse 1 (fly), we're also done after the special line
	if v == 1 {
		return strings.Join(lines, "\n")
	}
	
	// For other verses, add the cumulative "she swallowed X to catch Y" lines
	for i := v; i > 1; i-- {
		// Special case for the spider
		if i == 3 {
			lines = append(lines, fmt.Sprintf("She swallowed the %s to catch the %s that wriggled and jiggled and tickled inside her.", animals[i], animals[i-1]))
		} else {
			lines = append(lines, fmt.Sprintf("She swallowed the %s to catch the %s.", animals[i], animals[i-1]))
		}
	}
	
	// Last line is always the same for verses 2-7
	lines = append(lines, specialLines[1])
	
	return strings.Join(lines, "\n")
}

// Verses returns a string containing the verses from start to end, inclusive
func Verses(start, end int) string {
	var verses []string
	for i := start; i <= end; i++ {
		verses = append(verses, Verse(i))
	}
	return strings.Join(verses, "\n\n")
}

// Song returns the entire song (verses 1-8)
func Song() string {
	return Verses(1, 8)
}

File: javascript/food-chain/food-chain.js
export class Song {
  constructor() {
    this.animals = [
      'fly',
      'spider',
      'bird',
      'cat',
      'dog',
      'goat',
      'cow',
      'horse'
    ];
    
    this.secondLines = [
      'I don\'t know why she swallowed the fly. Perhaps she\'ll die.',
      'It wriggled and jiggled and tickled inside her.',
      'How absurd to swallow a bird!',
      'Imagine that, to swallow a cat!',
      'What a hog, to swallow a dog!',
      'Just opened her throat and swallowed a goat!',
      'I don\'t know how she swallowed a cow!',
      'She\'s dead, of course!'
    ];
  }

  verse(n) {
    // Adjust for 0-based indexing
    const index = n - 1;
    const animal = this.animals[index];
    
    // Special case for horse
    if (animal === 'horse') {
      return `I know an old lady who swallowed a horse.\nShe's dead, of course!\n`;
    }
    
    let verse = `I know an old lady who swallowed a ${animal}.\n`;
    verse += `${this.secondLines[index]}\n`;
    
    // Special case for fly - no chain of swallows
    if (animal === 'fly') {
      return verse;
    }
    
    // Add the chain of swallows in reverse order
    for (let i = index; i > 0; i--) {
      if (i === 2) { // Special case for bird -> spider
        verse += `She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n`;
      } else {
        verse += `She swallowed the ${this.animals[i]} to catch the ${this.animals[i-1]}.\n`;
      }
    }
    
    // Add the final line for all non-horse verses
    verse += `I don't know why she swallowed the fly. Perhaps she'll die.\n`;
    
    return verse;
  }

  verses(start, end) {
    let result = '';
    
    for (let i = start; i <= end; i++) {
      result += this.verse(i);
      
      // Add a blank line between verses
      if (i < end) {
        result += '\n';
      }
    }
    
    // Add an extra newline at the end for the expected format
    result += '\n';
    
    return result;
  }
}
